core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -155,27 +168,93 @@ public List<Snapshot> apply() {
     TableMetadata updated = internalApply();
     List<Snapshot> removed = Lists.newArrayList(base.snapshots());
     removed.removeAll(updated.snapshots());
-
     return removed;
   }
 
   private TableMetadata internalApply() {
     this.base = ops.refresh();
+    if (base.refs().isEmpty()) {
+      return base;
+    }
 
     Set<Long> idsToRetain = Sets.newHashSet();
-    List<Long> ancestorIds = SnapshotUtil.ancestorIds(base.currentSnapshot(), base::snapshot);
-    if (minNumSnapshots >= ancestorIds.size()) {
-      idsToRetain.addAll(ancestorIds);
-    } else {
-      idsToRetain.addAll(ancestorIds.subList(0, minNumSnapshots));
+    // Identify refs that should be removed
+    Map<String, SnapshotRef> retainedRefs = computeRetainedRefs(base.refs());
+    Set<Long> retainedRefIds = retainedRefs.entrySet().stream()
+        .map(refEntry -> refEntry.getValue().snapshotId())
+        .collect(Collectors.toSet());
+
+    for (long idToRemove : idsToRemove) {
+      Preconditions.checkArgument(!retainedRefIds.contains(idToRemove),
+          "Cannot expire %s. There is a reference to it", idToRemove);
```

Comment:
How about `"Cannot expire snapshot %s: referenced by %s %s", idToRemove, ref.isTag() ? "tag" : "branch", refName`?

Reply:
There can be multiple references for the given id; so to get this information the grouping from retained snapshot id to the list of refs has to be computed. I do think the extra computation is worth it, it is odd to to throw an exception "There is a reference to it" without mentioning what the references are. Will update. I think we can log the list of refs for the given id directly, I think in practice there would not be so many refs for a given snapshot id that it would pollute the logs to log it.

Reply:
I've updated to maintain the refs associated with a given snapshot ID so that the exception message could be clearer. @rdblue let me know what you think!